### ISSUE TITLE
DeviceClient doesnt dispose TCP connection if all method handlers are removed and no default handler specified

### ISSUE THREAD

**Issue by kristofferpagels** [Azure/azure-iot-sdk-csharp#879]
-**OS, version, SKU and CPU architecture used:** Windows 10 Desktop x64
-**Application's .NET Target Framework :** net472
-**Device:** Laptop
-**SDK version used:** 1.19.0
-**TransportType:** AMQP

## Description of the issue:
DeviceClient doesn't dispose the underlying TCP connection if we remove all method handlers before calling CloseAsync and don't have a method default handler specified. We expected it to close the tcp connection when calling closeAsync.

Below is a code sample exhibiting the issue. Two possible workarounds are described as well:

1. Not remove the last message handler. In this case it works as expected and the tcp connection is closed.
2. Add a method default handler. In this case it works as expected and the tcp connection is closed.

## Code sample exhibiting the issue:
```
    class Program
    {
        static async Task Main(string[] args)
        {
            string connectionString = "HostName=MyHub.azure-devices.net;DeviceId=mydevice;SharedAccessKey=****";
            try
            {
                using (var deviceClient = DeviceClient.CreateFromConnectionString(connectionString, TransportType.Amqp))
                {
                    await Open_NoMethodDefaultHandler_ReceiveFromDeviceClient_RemoveAllMethodHandlers_CloseDeviceClient_TcpConnectionIsNOTClosed(deviceClient);

                    // possible workarounds:
                    //await Open_NoMethodDefaultHandler_ReceiveFromDeviceClient_DoNotRemoveTheLastMethodHandler_CloseDeviceClient_TcpConnectionIsClosed(deviceClient);
                    //await Open_SetMethodDefaultHandler_ReceiveFromDeviceClient_RemoveAllMethodHandlers_CloseDeviceClient_TcpConnectionIsClosed(deviceClient);
                }

                Thread.Sleep(TimeSpan.FromMinutes(1));

                // At this point we expect the tcp connection count for the program to be zero.
            }
            catch (Exception e)
            {
                Console.WriteLine(e.Message);
            }

            Console.WriteLine("DONE");
            Console.ReadKey();
        }


        private static async Task Open_NoMethodDefaultHandler_ReceiveFromDeviceClient_RemoveAllMethodHandlers_CloseDeviceClient_TcpConnectionIsNOTClosed(DeviceClient deviceClient)
        {
            deviceClient.SetConnectionStatusChangesHandler(SomeConnectionStatusChangesHandler);

            await deviceClient.SetMethodHandlerAsync("someMethod1", SomeMethodCallback, new object());
            await deviceClient.SetMethodHandlerAsync("someMethod2", SomeMethodCallback, new object());

            var message = await deviceClient.ReceiveAsync(TimeSpan.FromSeconds(5));
            Console.WriteLine("Received message: " + message);

            await deviceClient.SetMethodHandlerAsync("someMethod1", null, null);
            // if we don't have a defaulthandler specified and remove the LAST methodhandler, the tcp connection will NOT be closed as expected when calling closeAsync.
            await deviceClient.SetMethodHandlerAsync("someMethod2", null, null);

            await deviceClient.CloseAsync();
        }

        private static async Task Open_NoMethodDefaultHandler_ReceiveFromDeviceClient_DoNotRemoveTheLastMethodHandler_CloseDeviceClient_TcpConnectionIsClosed(DeviceClient deviceClient)
        {
            deviceClient.SetConnectionStatusChangesHandler(SomeConnectionStatusChangesHandler);

            await deviceClient.SetMethodHandlerAsync("someMethod1", SomeMethodCallback, new object());
            await deviceClient.SetMethodHandlerAsync("someMethod2", SomeMethodCallback, new object());

            var message = await deviceClient.ReceiveAsync(TimeSpan.FromSeconds(5));
            Console.WriteLine("Received message: " + message);

            await deviceClient.SetMethodHandlerAsync("someMethod1", null, null);
            // If we don't remove the LAST methodhandler, the tcp connection will be closed as expected.
            //await deviceClient.SetMethodHandlerAsync("someMethod2", null, null);

            await deviceClient.CloseAsync();
        }

        private static async Task Open_SetMethodDefaultHandler_ReceiveFromDeviceClient_RemoveAllMethodHandlers_CloseDeviceClient_TcpConnectionIsClosed(DeviceClient deviceClient)
        {
            deviceClient.SetConnectionStatusChangesHandler(SomeConnectionStatusChangesHandler);

            await deviceClient.SetMethodHandlerAsync("someMethod1", SomeMethodCallback, new object());
            await deviceClient.SetMethodHandlerAsync("someMethod2", SomeMethodCallback, new object());
            await deviceClient.SetMethodDefaultHandlerAsync(MethodDefaultHandlerCallback, new object());

            var message = await deviceClient.ReceiveAsync(TimeSpan.FromSeconds(5));
            Console.WriteLine("Received message: " + message);

            await deviceClient.SetMethodHandlerAsync("someMethod1", null, null);
            // if we have a defaulthandler specified and remove the LAST methodhandler, the tcp connection will be closed as expected when calling closeAsync.
            await deviceClient.SetMethodHandlerAsync("someMethod2", null, null);

            await deviceClient.CloseAsync();
        }

        private static Task<MethodResponse> MethodDefaultHandlerCallback(MethodRequest methodRequest, object userContext)
        {
            Console.WriteLine("Received DirectMethodCall for unknown method");
            return Task.FromResult(new MethodResponse(501));
        }

        private static Task<MethodResponse> SomeMethodCallback(MethodRequest methodRequest, object userContext)
        {
            Console.WriteLine("Received DirectMethodCall");
            return Task.FromResult(new MethodResponse(200));
        }

        private static void SomeConnectionStatusChangesHandler(ConnectionStatus status, ConnectionStatusChangeReason reason)
        {
            Console.WriteLine("Status changed to: " + status);
        }
    }
```

**Comment by jenoola** [comment]
Got a repro for Amqp and Amqp_Tcp_Only but not for Amqp_WebSocket_Only. Investigating

**Comment by jenoola** [comment]
Issue fixed in the new release 1.20.0

**Comment by az-iot-builder-01** [comment]
@kristofferpagels, @jenoola, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)
